File: src/test/java/shopping/wish/service/WishServiceTest.java
package shopping.wish.service;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.util.ReflectionTestUtils;
import shopping.common.exception.ProductNotFoundException;
import shopping.common.exception.WishListNotFoundException;
import shopping.member.domain.Member;
import shopping.product.domain.Product;
import shopping.product.repository.ProductRepository;
import shopping.wish.domain.WishList;
import shopping.wish.domain.WishListItem;
import shopping.wish.dto.WishProductResponse;
import shopping.wish.repository.WishListItemRepository;
import shopping.wish.repository.WishListRepository;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

@ExtendWith(MockitoExtension.class)
public class WishServiceTest {

    @Mock
    private WishListRepository wishListRepository;

    @Mock
    private WishListItemRepository wishListItemRepository;

    @Mock
    private ProductRepository productRepository;

    @InjectMocks
    private WishService wishService;

    @Test
    void 위시리스트에_상품을_추가한다() {

        //given
        Long memberId = 1L;
        Long productId = 1L;

        WishList wishList = mock(WishList.class);
        Product product = mock(Product.class);

        given(wishListRepository.findByMemberId(memberId))
                .willReturn(Optional.of(wishList));

        given(productRepository.findById(productId))
                .willReturn(Optional.of(product));

        //when
        wishService.addWish(memberId, productId);

        //then
        verify(wishListItemRepository).save(any(WishListItem.class));
    }

    @Test
    void 위시리스트_아이템을_삭제한다() {

        Long wishId = 1L;

        wishService.deleteWish(wishId);

        verify(wishListItemRepository).deleteById(wishId);
    }

    @Test
    void 위시리스트_아이템_목록을_조회한다() {

        //given
        Long memberId = 1L;
        Member member = Member.create("test@test.com", "123456");

        WishList wishList = WishList.create(member);

        ReflectionTestUtils.setField(wishList, "id", 1L);

        Product product = new Product("아이폰", 1, 1_000_000, "image.jpg");
        ReflectionTestUtils.setField(product, "id", 1L);

        WishListItem item = WishListItem.create(wishList, product);

        given(wishListRepository.findByMemberId(memberId))
                .willReturn(Optional.of(wishList));

        given(wishListItemRepository.findByWishListId(1L))
                .willReturn(List.of(item));

        //when
        List<WishProductResponse> result = wishService.getWishList(memberId);

        //then
        assertThat(result).hasSize(1);
        assertThat(result.getFirst().getProductId()).isEqualTo(1L);
    }

    @Test
    void 위시리스트가_없으면_예외가_발생한다() {
        //given
        Long memberId = 1L;

        //when & then
        assertThrows(WishListNotFoundException.class, () -> wishService.getWishList(memberId));
    }

    @Test
    void 존재하지_않는_상품이면_예외가_발생한다() {
        // given
        Long productId = 999L;

        // when & then
        assertThrows(ProductNotFoundException.class, () -> wishService.addWish(1L, productId));
    }
}

File: src/test/kotlin/shopping/application/wish/WishServiceTest.kt
package shopping.application.wish

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.instancio.Instancio
import shopping.domain.product.ProductRepository
import shopping.domain.wish.Wish
import shopping.domain.wish.WishRepository
import shopping.support.error.CoreException

class WishServiceTest :
    BehaviorSpec({

        val wishRepository = mockk<WishRepository>()
        val productRepository = mockk<ProductRepository>()
        val wishService = WishService(wishRepository, productRepository)

        given("위시리스트에 상품을 추가할 때") {

            val memberId = 1L
            val productId = 100L

            `when`("정상적인 요청이면") {
                val wish = Wish(memberId, productId)

                every { productRepository.existsById(productId) } returns true
                every { wishRepository.existsByMemberIdAndProductId(memberId, productId) } returns false
                every { wishRepository.save(any()) } returns wish

                then("위시리스트에 저장된다") {
                    val result = wishService.addWish(memberId, productId)
                    result.memberId shouldBe memberId
                    result.productId shouldBe productId
                    verify { wishRepository.save(any()) }
                }
            }

            `when`("존재하지 않는 상품이면") {
                every { productRepository.existsById(productId) } returns false

                then("예외가 발생한다") {
                    val exception =
                        shouldThrow<CoreException> {
                            wishService.addWish(memberId, productId)
                        }
                    exception.message shouldBe "상품이 존재하지 않습니다."
                }
            }

            `when`("이미 위시리스트에 있는 상품이면") {
                every { productRepository.existsById(productId) } returns true
                every { wishRepository.existsByMemberIdAndProductId(memberId, productId) } returns true

                then("예외가 발생한다") {
                    val exception =
                        shouldThrow<CoreException> {
                            wishService.addWish(memberId, productId)
                        }
                    exception.message shouldBe "이미 위시리스트에 추가된 상품입니다."
                }
            }
        }

        given("위시리스트에서 상품을 제거할 때") {

            val memberId = 1L
            val wishId = 50L

            `when`("정상적으로 제거 요청하면") {
                // ID를 직접 설정하는 대신 Instancio가 생성한 객체를 사용
                val wish = Instancio.create(Wish::class.java)

                every { wishRepository.findByMemberId(memberId) } returns listOf(wish)
                every { wishRepository.deleteById(any()) } returns Unit

                then("상품이 위시리스트에서 제거된다") {
                    wishService.removeWish(memberId, wish.id)
                    verify { wishRepository.deleteById(wish.id) }
                }
            }

            `when`("본인의 위시리스트 항목이 아니면") {
                every { wishRepository.findByMemberId(memberId) } returns emptyList()

                then("예외가 발생한다") {
                    val exception =
                        shouldThrow<CoreException> {
                            wishService.removeWish(memberId, wishId)
                        }
                    exception.message shouldBe "위시리스트에 존재하지 않는 항목입니다."
                }
            }
        }

        given("위시리스트 목록을 조회할 때") {

            val memberId = 1L

            `when`("조회 요청하면") {
                // memberId를 생성자로 직접 주입하여 정확성 보장
                val wishes = List(3) { Wish(memberId, it.toLong()) }

                every { wishRepository.findByMemberId(memberId) } returns wishes

                then("사용자의 위시리스트 목록을 반환한다") {
                    val result = wishService.getWishes(memberId)
                    result.size shouldBe 3
                    result.all { it.memberId == memberId } shouldBe true
                }
            }
        }
    })
